nvflare/fuel/hci/binary_proto.py

```diff
@@ -0,0 +1,454 @@
+        # If the value is not BINARY_MARKER, then it is treated as text protocol. This is to be backward compatible
+        # with the current text-based protocol!
+        data = self.receiver.recv(1)
+        if not data:
+            raise RuntimeError("no data type marker received")
+
+        marker = data[0]
+        if marker == BINARY_MARKER:
+            # Binary protocol - process according to binary protocol definition
+            self.content_type = CT_BINARY
+            buffer = self._must_recv(HEADER_LEN)
+            if not buffer:
+                raise RuntimeError(f"cannot get header buffer of {HEADER_LEN} bytes")
+            if len(buffer) != HEADER_LEN:
+                raise RuntimeError(f"expect {HEADER_LEN} header bytes but only got {len(buffer)}")
+
+            meta_size, body_size = HEADER_STRUCT.unpack_from(buffer, 0)
+            if meta_size < 0:
+                raise RuntimeError(f"invalid binary data meta size {meta_size}")
+
+            # get meta
+            meta_bytes = self._must_recv(meta_size)
+            if not data:
+                raise RuntimeError("no meta data received")
+            if len(meta_bytes) != meta_size:
+                raise RuntimeError(f"expect {meta_size} meta bytes but got {len(meta_bytes)}")
+
+            # meta data must be str!
+            self.meta = str(meta_bytes, "utf-8")
+            self._parse_binary(body_size)
+        else:
+            # text content - the 1st byte is part of the data!
+            self.content_type = CT_TEXT
+            self.processor.process(data, CT_TEXT)
+            self._parse_text()
+
+        self.processor.finalize()
+
+
+class MsgDataProcessor(DataProcessor):
+    """The MsgDataProcessor is a special DataProcessor that can handle both TEXT and BINARY content type.
+    For TEXT, it collects all received text segments and assemble them into one text string;
+    For BINARY, it saves received data into a temporary file so that no memory is used to collect the data. This
+    is necessary to support extremely large data exchanges (e.g. large job submission and download).
+
+    """
+
+    def __init__(self):
+        self.file = None
+        self.file_name = None
+        self.text_segs = []
+        self.total_text = None
+
+    def process(self, data, content_type: int):
+        if content_type == CT_TEXT:
+            data = str(data, "utf-8")
+            end_idx = data.find(ALL_END)
+            if end_idx >= 0:
+                self.text_segs.append(data[:end_idx])
+                return True  # all received
+            else:
+                self.text_segs.append(data)
+        else:
+            # binary - write to file
+            if not self.file:
+                self.file_name = os.path.join(tempfile.gettempdir(), str(uuid.uuid4()))
+                self.file = open(self.file_name, "w+b")
+            self.file.write(data)
```

Comment:
This should be very rare since file is on temp dir and randomly named. But this could happen if someone messed with the file system. In this case, we'll just let exception happen.